src/sage/schemes/elliptic_curves/cm.py

```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
+        orders of each class number `h\le100`.  It is easy to extend
+        this to larger class number conditional on the GRH, but much
+        harder to obyain unconditional results.
+
+    INPUT:
+
+    - `h` -- integer
+
+    EXAMPLES::
+
+        sage: from sage.schemes.elliptic_curves.cm import largest_disc_with_class_number
+        sage: largest_disc_with_class_number(0)
+        (0, 0)
+        sage: largest_disc_with_class_number(1)
+        (163, 13)
+        sage: largest_disc_with_class_number(2)
+        (427, 29)
+        sage: largest_disc_with_class_number(10)
+        (13843, 123)
+        sage: largest_disc_with_class_number(100)
+        (1856563, 2311)
+        sage: largest_disc_with_class_number(101)
+        Traceback (most recent call last):
+        ...
+        NotImplementedError: largest discriminant not available for class number 101
+
+    For most `h\le100`, the largest fundamental discriminant with class number `h` is also the largest discriminant, but this is not the case for `h=`::
+
+        sage: from sage.schemes.elliptic_curves.cm import largest_disc_with_class_number, largest_fundamental_disc_with_class_number
+        sage: [h for h in range(1,101) if largest_disc_with_class_number(h)[0] != largest_fundamental_disc_with_class_number(h)[0]]
+        [6, 8, 12, 16, 20, 30, 40, 42, 52, 70]
+        sage: largest_fundamental_disc_with_class_number(6)
+        (3763, 51)
+        sage: largest_disc_with_class_number(6)
+        (4075, 101)
+    """
+    h = Integer(h)
+    if h <= 0:
+        return Integer(0), Integer(0)
+    try:
+        B, c = klaise_table[h]
+        return (Integer(B), Integer(c))
+    except KeyError:
+        raise NotImplementedError("largest discriminant not available for class number %s" % h)
+
+# This dict has class numbers h as keys, the value at h is a complete
+# list of paits (D0,f) such that D=D0*f**2 has class number h.  We
+# initialise it with h=1 only; other values will be added by calls to
+# discriminants_with_bounded_class_number().
 
+hDf_dict = {ZZ(1): [(ZZ(D), ZZ(h)) for D,h in
+                    [(-3, 1), (-3, 2), (-3, 3), (-4, 1), (-4, 2), (-7, 1), (-7, 2),
+                     (-8, 1), (-11, 1), (-19, 1), (-43, 1), (-67, 1), (-163, 1)]]}
 
 @cached_function
 def discriminants_with_bounded_class_number(hmax, B=None, proof=None):
     r"""
-    Return dictionary with keys class numbers `h\le hmax` and values the
-    list of all pairs `(D, f)`, with `D<0` a fundamental discriminant such
-    that `Df^2` has class number `h`.  If the optional bound `B` is given,
-    return only those pairs with fundamental `|D| \le B`, though `f` can
-    still be arbitrarily large.
+    Return a dictionary with keys class numbers `h\le hmax` and values the
+    list of all pairs `(D_0, f)`, with `D_0<0` a fundamental discriminant such
+    that `D=D_0f^2` has class number `h`.  If the optional bound `B` is given,
+    return only those pairs with fundamental `|D| \le B`.
```

Comment:
Right, I deleted "fundamental".   The previous behaviour was unnatural and only implemented that way (by WAS and/or me) because with the previous algorithm that was all you could do. *Please* don't insist on a deprecation!  If you look at the old doctest output it was quite confusing.

This function is not referred to outside this file and could have been hidden (preceding underscore), as it is only called by ``cm_orders`` and ``cm_j_invariants_and_orders``, but I don't want to do that now as it really would require a deprecation.